File: 2.3 correccion gamma/correccionGamma.py
# ///////////////////////
# Correccion gama
# Autor: Bryan Medina
# ///////////////////////

import cv2
import numpy as np
import matplotlib.pyplot as plt

# ubicacion de la imagen
f = cv2.imread('sigure.jpg', 0)

#asignamos gamma y obtenemos las constantes c encargadas de mantener a raya los valores del despliegue (0 - 255)
# en este caso con y obtenemos la función que graficaremos.
gamma = 0.4
x = np.array(range(256))
c = 255/np.power(255, gamma)
y =  c * np.power(x, gamma)

gamma1 = 1.0
c1 = 255/np.power(255, gamma1)
y1 =  c1 * np.power(x, gamma1)

gamma2 = 2.5
c2 = 255/np.power(255, gamma2)
y2 =  c2 * np.power(x, gamma2)

#graficamos las funciones

plt.figure()
plt.title('correción gamma')
plt.plot(x, y, label = '$\gamma = 0.4$')
plt.plot(x, y1, label = '$\gamma = 1.0$')
plt.plot(x, y2, label = '$\gamma = 2.5$')
plt.legend(loc='upper left')
plt.xlabel('intensidades de entrada')
plt.ylabel('intensidades de salida')
plt.show()

#calculamos la transformacion gamma, con la formulita
outputImg = np.array(c * np.power(f, gamma), dtype = 'uint8')
outputImg1 = np.array(c2 * np.power(f, gamma2), dtype = 'uint8')


cv2.imshow('gamma = 0.5', outputImg)
cv2.imshow('gamma = 2.0', outputImg1)
cv2.imshow('entrada', f)
#cv2.imwrite('punto5.jpg', outputImg)
#cv2.imwrite('2puntocero.jpg', outputImg1)
cv2.waitKey(0)
cv2.destroyAllWindows()

File: 2.3 correccion gamma/correccionGamma.m
%////////////////////////////
%/correccion gamma
%//Autor: Bryan Medina
%////////////////////////////

f = imread('sigure.jpg');
f = rgb2gray(f);

gamma = 0.2;
gamma1 = 5;

%constante  para poner en el mismo rango de la imagen

c = 255/(255^gamma);
c1 = 255/(255^gamma1);

% primero se realiza un escalamiento de la intensidad de la imagen de 0 a 1
% donde estos valores despues son mapeados a los 256 valores de intensidad
g = im2uint8(mat2gray(c * double(f).^gamma));
g1 = im2uint8(mat2gray(c1 * double(f).^gamma1));


figure;
subplot(1, 3, 1);
imshow(f);
title('gamma = 1');

subplot(1, 3, 2);
imshow(g);
title('gamma = 0.2');

subplot(1, 3, 3)
imshow(g1);
title('gamma = 5');
